File: baekjoon/baekjoon_14719.py
import sys

H, W = map(int, sys.stdin.readline().split())
ground = list(map(int, sys.stdin.readline().split()))

left_point, right_point = 0, W - 1
left_max, right_max = ground[left_point], ground[right_point]
rain = 0
while left_point < right_point: # 두 포인터가 만나면 종료
    if left_max < right_max:    # 왼쪽의 최대 높이가 오른쪽의 최대 높이보다 낮다면
        left_point += 1         # 왼쪽 포인터 증가
        left_max = max(left_max, ground[left_point]) # 최대값 갱신하고
        if left_max > ground[left_point]:            # 포인터가 바라보는 위치의 블록 높이가 최대 높이보다 낮다면
            rain += left_max - ground[left_point]    # 빗물 카운트

    else:                       # 오른쪽의 최대 높이가 왼쪽의 최대 높이보다 낮다면
        right_point -= 1        # 이하 동일
        right_max = max(right_max, ground[right_point])
        if right_max > ground[right_point]:
            rain += right_max - ground[right_point]

print(rain)

File: baekjoon/baekjoon14719.java
package baekjoon;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.IOException;
import java.util.StringTokenizer;

public class baekjoon14719 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringBuilder sb = new StringBuilder();

        StringTokenizer st = new StringTokenizer(br.readLine());
        int H = Integer.parseInt(st.nextToken());
        int W = Integer.parseInt(st.nextToken());

        int[] ground = new int[W];
        st = new StringTokenizer(br.readLine());
        for (int w = 0; w < W; w++) {
            ground[w] = Integer.parseInt(st.nextToken());
        }

        int leftPointer = 0, rightPointer = W - 1, rain = 0;
        int leftMax = ground[leftPointer], rightMax = ground[rightPointer];

        while (leftPointer < rightPointer) {
            if (leftMax < rightMax) {
                leftMax = Math.max(leftMax, ground[++leftPointer]);
                if (leftMax > ground[leftPointer]) {
                    rain += leftMax - ground[leftPointer];
                }
            } else {
                rightMax = Math.max(rightMax, ground[--rightPointer]);
                if (rightMax > ground[rightPointer]) {
                    rain += rightMax - ground[rightPointer];
                }
            }
        }

        sb.append(rain);
        bw.write(sb.toString());
        bw.flush();
        sb.setLength(0);

        br.close();
        bw.close();
    }
}
